scripts/generate_markdowns.py

```diff
@@ -1,203 +1,60 @@
-
-md_exporter = MarkdownExporter(exclude_output=True)
-
-
-def get_notebook_number(nb_path):
-    nb_filename = nb_path.stem
-    return int(re.search("\d+", nb_filename.split("_")[0]).group(0))
-
-
-def generate_markdown_from_notebook(nb_path):
-    body, _ = md_exporter.from_filename(nb_path)
-    n = get_notebook_number(nb_path)
-    print(f"Processing {nb_path}")
-
-    with open(markdown_tutorials_dir / f"{n}.md", "w", encoding="utf-8") as f:
+def read_index(path):
+    with open(path, "rb") as f:
+        return tomli.load(f)
+
+
+def generate_frontmatter(config, tutorial):
+    aliases = []
+    if "aliases" in tutorial:
+        for alias in tutorial["aliases"]:
+            aliases.append(f"/tutorials/{alias}")
+
+    frontmatter = f"""---
+layout: {config["layout"]}
+colab: {config["colab"]}{tutorial["notebook"]}
+toc: {config["toc"]}
+title: "{tutorial["title"]}"
+last_updated: {date.today()}
+level: "{tutorial["level"]}"
+weight: {tutorial["weight"]}
+description: {tutorial["description"]}
+category: "QA"
+aliases: {aliases}
+---
+    """
+    return frontmatter
+
+
+def generate_markdown_from_notebook(config, tutorial, output_path, tutorials_path):
+    frontmatter = generate_frontmatter(config, tutorial)
+    md_exporter = MarkdownExporter(exclude_output=True)
+    body, _ = md_exporter.from_filename(f"{tutorials_path}")
+    print(f"Processing {tutorials_path}")
+
+    with open(f"{output_path}/{tutorial['notebook'][:-6]}.md", "w", encoding="utf-8") as f:
         try:
-            f.write(headers[n] + "\n\n")
+            f.write(frontmatter + "\n\n")
         except IndexError as e:
             raise IndexError(
                 "Can't find the header for this tutorial. Have you added it in 'scripts/generate_markdowns.py'?"
             ) from e
         f.write(body)
 
 
-def main(argv: Sequence[str] = sys.argv):
-    parser = argparse.ArgumentParser()
-    parser.add_argument("filenames", nargs="*", help="Notebooks filenames")
-    parser.add_argument("--all", action="store_true", help="Generate markdown version for all the notebooks")
-    args = parser.parse_args(argv)
-
-    filenames = args.filenames
-    if args.all:
-        filenames = notebook_tutorials_dir.glob("*.ipynb")
-
-    for filename in filenames:
-        filepath = Path(filename).resolve()
-        if filepath.parent == notebook_tutorials_dir and filepath.suffix == ".ipynb":
-            generate_markdown_from_notebook(filepath)
-
-    return 0
-
-
 if __name__ == "__main__":
-    raise SystemExit(main())
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--index", dest="index")
+    parser.add_argument("--notebooks", dest="notebooks", nargs='+', default=[])
+    parser.add_argument("--output", dest="output", default="markdowns")
+    args = parser.parse_args()
+    index = read_index(args.index)
+
+    for notebook in args.notebooks:
+        for tutorial in index["tutorial"]:
+            if  tutorial["notebook"] in notebook:
```

Comment:
I had the impression offline that you weren't happy with the nested `for`s. One solution might be creating a map pointing each notebook name to the tutorial configuration dict from the index:

nb_name -> tutorial_dict_containing_nb

This way you would cycle the index only once to build the map, then you can access it how many times you want in O(1) using the notebook name.

The code would look like this:
```
    ...
    index = read_index(args.index)
    # map the notebook name to the corresponding tutorial
    nb_to_tutorial_cfg = {cfg["notebook"]: cfg for cfg in index["tutorial"]}

    for notebook in args.notebooks:
        # extract the notebook name from whatever path was passed to the script
        nb_name = Path(notebook).name
        tutorial_cfg = nb_to_tutorial_cfg.get(nb_name)
        # convert the tutorial if the notebook is valid
        if tutorial_cfg:
            generate_markdown_from_notebook(index["config"], tutorial_cfg, args.output, notebook)
```